File: app/src/main/java/com/example/app/MainActivity.java
package com.example.app;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.ImageView;
import android.widget.Toast;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.view.View;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;

public class MainActivity extends Activity {
    boolean DoublePressToExit = false;
    private boolean isMuted = false; // Initial mute state
    Toast toast;
    // creating a variable for
    // button and media player

    ImageView playbtn;
    ImageView btnMute;

    private MediaPlaybackService mediaPlaybackService;
    private boolean isBound = false;

    Handler handler;

    private WebView mWebView;

    @Override
    @SuppressLint("SetJavaScriptEnabled")
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mWebView = findViewById(R.id.activity_main_webview);
        WebSettings webSettings = mWebView.getSettings();
        webSettings.setJavaScriptEnabled(true);
        mWebView.setWebViewClient(new MyWebViewClient());
        mWebView.addJavascriptInterface(new WebAppInterface(this), "Android");

        // REMOTE RESOURCE
        mWebView.loadUrl("https://purplescorpion1.github.io/");

        // LOCAL RESOURCE
        // mWebView.loadUrl("file:///android_asset/index.html");

        playbtn = findViewById(R.id.btnplay);
        btnMute = findViewById(R.id.btnmute);

        playbtn.setOnClickListener(btnClickListen);

        handler = new Handler();

        // Find your ImageView (btnmute) by its ID
        ImageView btnMute = findViewById(R.id.btnmute);

        // Set an OnClickListener for the ImageView
        btnMute.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                toggleMute(); // Toggle mute state when the image is clicked
            }
        });

        Intent intent = new Intent(this, MediaPlaybackService.class);
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (isBound) {
            unbindService(serviceConnection);
            isBound = false;
        }
    }

    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            MediaPlaybackService.LocalBinder binder = (MediaPlaybackService.LocalBinder) service;
            mediaPlaybackService = binder.getService();
            isBound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            isBound = false;
        }
    };


    private View.OnClickListener btnClickListen = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            if (isBound) {
                if (mediaPlaybackService.isPlaying()) {
                    mediaPlaybackService.stop();
                    playbtn.setImageResource(R.drawable.play);
                    Toast.makeText(MainActivity.this, "Audio has been paused", Toast.LENGTH_SHORT).show();
                } else {
                    mediaPlaybackService.play();
                    playbtn.setImageResource(R.drawable.pause);
                    Toast.makeText(MainActivity.this, "Audio is starting. Please wait...", Toast.LENGTH_SHORT).show();
                }
            }
        }
    };

    private void toggleMute() {
        if (isBound && mediaPlaybackService.isPlaying()) {
            if (isMuted) {
                mediaPlaybackService.unmute();
                isMuted = false;
                btnMute.setImageResource(R.drawable.mute);
                Toast.makeText(MainActivity.this, "Audio has been unmuted", Toast.LENGTH_SHORT).show();
            } else {
                mediaPlaybackService.mute();
                isMuted = true;
                btnMute.setImageResource(R.drawable.unmute);
                Toast.makeText(MainActivity.this, "Audio has been muted", Toast.LENGTH_SHORT).show();
            }
        }
    }

    @Override
    public void onBackPressed() {
        if (DoublePressToExit){
            finish();
            toast.cancel();
        }else {
            DoublePressToExit=true;
            Toast.makeText(this, "Press Again To Exit", Toast.LENGTH_SHORT).show();
            Handler handler=new Handler(Looper.getMainLooper());
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    DoublePressToExit=false;
                }
            },2000);
        }
    }

    public class WebAppInterface {
        Context mContext;

        WebAppInterface(Context c) {
            mContext = c;
        }

        @android.webkit.JavascriptInterface
        public void showToast(String toast) {
            Toast.makeText(mContext, toast, Toast.LENGTH_SHORT).show();
        }

        @android.webkit.JavascriptInterface
        public void updateTrackInfo(String title, String imageUrl) {
            if (isBound) {
                new DownloadImageTask().execute(imageUrl, title);
            }
        }
    }

    private class DownloadImageTask extends android.os.AsyncTask<String, Void, android.graphics.Bitmap> {
        private String title;

        @Override
        protected android.graphics.Bitmap doInBackground(String... urls) {
            String imageUrl = urls[0];
            title = urls[1];
            android.graphics.Bitmap bitmap = null;
            try {
                java.io.InputStream in = new java.net.URL(imageUrl).openStream();
                bitmap = android.graphics.BitmapFactory.decodeStream(in);
            } catch (Exception e) {
                e.printStackTrace();
            }
            return bitmap;
        }

        @Override
        protected void onPostExecute(android.graphics.Bitmap result) {
            if (isBound) {
                mediaPlaybackService.updateNotification(title, result);
            }
        }
    }
}

File: app/src/main/java/com/bootiemashup/radio/MainActivity.kt
package com.bootiemashup.radio

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.media.AudioManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.KeyEvent
import android.view.View
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.media3.common.MediaMetadata
import androidx.media3.common.Player
import androidx.media3.session.MediaController
import androidx.media3.session.SessionToken
import androidx.mediarouter.app.MediaRouteButton
import androidx.mediarouter.media.MediaControlIntent
import androidx.mediarouter.media.MediaRouteSelector
import com.bumptech.glide.Glide
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.MoreExecutors
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.Request
import org.json.JSONObject

class MainActivity : AppCompatActivity() {

    private var mediaControllerFuture: ListenableFuture<MediaController>? = null
    private var mediaController: MediaController? = null

    private lateinit var ivArtwork: ImageView
    private lateinit var tvTrackTitle: TextView
    private lateinit var tvNextTrackLabel: TextView
    private lateinit var tvNextTrack: TextView
    private lateinit var btnPlayPause: ImageButton
    private lateinit var btnMute: ImageButton
    private lateinit var mediaRouteButton: MediaRouteButton

    private var doubleBackToExitPressedOnce = false
    private var uiPollingJob: kotlinx.coroutines.Job? = null
    private var lastUiNowPlaying: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Bind Views
        ivArtwork = findViewById(R.id.ivArtwork)
        tvTrackTitle = findViewById(R.id.tvTrackTitle)
        tvNextTrackLabel = findViewById(R.id.tvNextTrackLabel)
        tvNextTrack = findViewById(R.id.tvNextTrack)
        btnPlayPause = findViewById(R.id.btnPlayPause)
        btnMute = findViewById(R.id.btnMute)
        mediaRouteButton = findViewById(R.id.mediaRouteButton)

        // Configure MediaRouteButton for standard Bluetooth / audio routing
        val mediaRouteSelector = MediaRouteSelector.Builder()
            .addControlCategory(MediaControlIntent.CATEGORY_LIVE_AUDIO)
            .addControlCategory(MediaControlIntent.CATEGORY_REMOTE_PLAYBACK)
            .build()
        mediaRouteButton.routeSelector = mediaRouteSelector

        // Setup double back press to exit
        val onBackPressedCallback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (doubleBackToExitPressedOnce) {
                    killAudioAndExit()
                } else {
                    doubleBackToExitPressedOnce = true
                    Toast.makeText(this@MainActivity, "Press back again to exit", Toast.LENGTH_SHORT).show()
                    Handler(Looper.getMainLooper()).postDelayed({
                        doubleBackToExitPressedOnce = false
                    }, 2000)
                }
            }
        }
        onBackPressedDispatcher.addCallback(this, onBackPressedCallback)

        // Support Android TV focus and D-pad navigation
        if (isAndroidTV()) {
            btnPlayPause.isFocusable = true
            btnMute.isFocusable = true
            mediaRouteButton.isFocusable = true

            // Set up clean background coloring or focus change listeners to indicate focus
            val focusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
                if (hasFocus) {
                    v.animate().scaleX(1.15f).scaleY(1.15f).setDuration(150).start()
                } else {
                    v.animate().scaleX(1.0f).scaleY(1.0f).setDuration(150).start()
                }
            }
            btnPlayPause.onFocusChangeListener = focusChangeListener
            btnMute.onFocusChangeListener = focusChangeListener
            mediaRouteButton.onFocusChangeListener = focusChangeListener

            btnPlayPause.requestFocus()
        }

        // Setup button listeners
        btnPlayPause.setOnClickListener {
            togglePlayback()
        }

        btnMute.setOnClickListener {
            toggleMute()
        }
    }

    override fun onStart() {
        super.onStart()
        // Start foreground service and connect to Media3 Session
        try {
            val serviceIntent = Intent(this, PlaybackService::class.java)
            ContextCompat.startForegroundService(this, serviceIntent)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        connectToMediaSession()
    }

    override fun onResume() {
        super.onResume()
        startUiPolling()
    }

    override fun onPause() {
        super.onPause()
        stopUiPolling()
    }

    override fun onStop() {
        super.onStop()
        stopUiPolling()
        // Release MediaController connection
        releaseMediaController()
    }

    private fun connectToMediaSession() {
        val sessionToken = SessionToken(this, ComponentName(this, PlaybackService::class.java))
        mediaControllerFuture = MediaController.Builder(this, sessionToken).buildAsync()
        mediaControllerFuture?.addListener({
            try {
                val controller = mediaControllerFuture?.get()
                if (controller != null) {
                    onMediaControllerConnected(controller)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }, ContextCompat.getMainExecutor(this))
    }

    private fun onMediaControllerConnected(controller: MediaController) {
        mediaController = controller

        // Listen for playback and metadata changes from session
        controller.addListener(object : Player.Listener {
            override fun onIsPlayingChanged(isPlaying: Boolean) {
                updatePlayPauseUI(controller.playWhenReady)
            }

            override fun onPlayWhenReadyChanged(playWhenReady: Boolean, reason: Int) {
                updatePlayPauseUI(playWhenReady)
            }

            override fun onPlaybackStateChanged(playbackState: Int) {
                updatePlayPauseUI(controller.playWhenReady)
            }

            override fun onVolumeChanged(volume: Float) {
                updateMuteUI(volume == 0f)
            }

            override fun onPlaylistMetadataChanged(metadata: MediaMetadata) {
                updateTrackMetadataUI(metadata)
            }

            override fun onMediaMetadataChanged(metadata: MediaMetadata) {
                updateTrackMetadataUI(controller.playlistMetadata)
            }
        })

        // Initial UI sync
        updatePlayPauseUI(controller.playWhenReady)
        updateMuteUI(controller.volume == 0f)
        updateTrackMetadataUI(controller.playlistMetadata)
    }

    private fun releaseMediaController() {
        mediaControllerFuture?.let {
            MediaController.releaseFuture(it)
            mediaControllerFuture = null
        }
        mediaController = null
    }

    private fun togglePlayback() {
        val controller = mediaController ?: return
        if (controller.playWhenReady) {
            controller.pause()
        } else {
            if (controller.playbackState == Player.STATE_IDLE || controller.playbackState == Player.STATE_ENDED) {
                controller.prepare()
            }
            controller.play()
        }
        // Also send custom action to service for complete sync
        try {
            controller.sendCustomCommand(
                androidx.media3.session.SessionCommand("ACTION_TOGGLE_PLAY_PAUSE", Bundle.EMPTY),
                Bundle.EMPTY
            )
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun toggleMute() {
        val controller = mediaController ?: return
        val currentVolume = controller.volume
        if (currentVolume == 0f) {
            controller.volume = 1f
            updateMuteUI(false)
        } else {
            controller.volume = 0f
            updateMuteUI(true)
        }
        // Also trigger service custom action for complete sync
        try {
            controller.sendCustomCommand(
                androidx.media3.session.SessionCommand("ACTION_TOGGLE_MUTE", Bundle.EMPTY),
                Bundle.EMPTY
            )
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun updatePlayPauseUI(isPlaying: Boolean) {
        if (isPlaying) {
            btnPlayPause.setImageResource(R.drawable.ic_pause_white_24dp)
            btnPlayPause.contentDescription = "Pause"
        } else {
            btnPlayPause.setImageResource(R.drawable.ic_play_arrow_white_24dp)
            btnPlayPause.contentDescription = "Play"
        }
    }

    private fun updateMuteUI(isMuted: Boolean) {
        if (isMuted) {
            btnMute.setImageResource(R.drawable.ic_volume_off_white_24dp)
            btnMute.contentDescription = "Unmute Audio"
        } else {
            btnMute.setImageResource(R.drawable.ic_volume_up_white_24dp)
            btnMute.contentDescription = "Mute Audio"
        }
    }

    private fun updateTrackMetadataUI(metadata: MediaMetadata?) {
        val activeMetadata = if (metadata != null && (!metadata.displayTitle.isNullOrEmpty() || !metadata.title.isNullOrEmpty())) {
            metadata
        } else {
            mediaController?.playlistMetadata ?: metadata
        } ?: return

        // Update Track Title
        val nowPlaying = activeMetadata.displayTitle?.toString() ?: ""
        val artist = activeMetadata.artist?.toString() ?: ""
        val title = activeMetadata.title?.toString() ?: ""

        val displayText = when {
            nowPlaying.isNotEmpty() -> nowPlaying
            artist.isNotEmpty() && title.isNotEmpty() -> "$artist - $title"
            title.isNotEmpty() -> title
            else -> "Bootie Mashup Radio"
        }
        tvTrackTitle.text = displayText

        // Update Coming Next Track Info
        val nextTrack = activeMetadata.extras?.getString("next_track") ?: ""
        if (nextTrack.isNotEmpty()) {
            tvNextTrackLabel.visibility = View.VISIBLE
            tvNextTrack.visibility = View.VISIBLE
            tvNextTrack.text = nextTrack
        } else {
            tvNextTrackLabel.visibility = View.GONE
            tvNextTrack.visibility = View.GONE
        }

        // Update Album Artwork using Glide without placeholder flashing
        val artworkUri = activeMetadata.artworkUri

        val requestBuilder = Glide.with(this@MainActivity)
            .load(artworkUri ?: "https://c7.radioboss.fm/w/artwork/205.jpg")
            .error(R.mipmap.ic_launcher_round)

        if (ivArtwork.drawable != null) {
            requestBuilder.placeholder(ivArtwork.drawable)
        } else {
            requestBuilder.placeholder(R.mipmap.ic_launcher_round)
        }

        requestBuilder.into(ivArtwork)
    }

    private fun startUiPolling() {
        uiPollingJob?.cancel()
        uiPollingJob = CoroutineScope(Dispatchers.IO).launch {
            while (isActive) {
                try {
                    fetchAndUpdateUiMetadata()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                delay(5000)
            }
        }
    }

    private fun stopUiPolling() {
        uiPollingJob?.cancel()
        uiPollingJob = null
    }

    private suspend fun fetchAndUpdateUiMetadata() {
        val request = Request.Builder()
            .url("https://c7.radioboss.fm/w/nowplayinginfo?u=205")
            .build()

        PlaybackService.okHttpClient.newCall(request).execute().use { response ->
            if (response.isSuccessful) {
                val jsonStr = response.body?.string() ?: ""
                val json = JSONObject(jsonStr)
                val nowPlaying = json.optString("nowplaying", "").trim()
                var artist = json.optString("currenttrack_artist", "").trim()
                var title = json.optString("currenttrack_title", "").trim()
                val nextTrack = json.optString("nexttrack", "").trim()

                if (nowPlaying.isNotBlank() && nowPlaying == lastUiNowPlaying) {
                    return@use
                }

                if (nowPlaying.isNotBlank()) {
                    lastUiNowPlaying = nowPlaying
                }

                if (artist.isBlank() || title.isBlank()) {
                    if (nowPlaying.contains(" - ")) {
                        val parts = nowPlaying.split(" - ", limit = 2)
                        if (artist.isBlank()) artist = parts[0].trim()
                        if (title.isBlank()) title = parts[1].trim()
                    } else {
                        if (title.isBlank()) title = if (nowPlaying.isNotBlank()) nowPlaying else "Live Stream"
                        if (artist.isBlank()) artist = "Bootie Mashup Radio"
                    }
                }

                val displayText = if (nowPlaying.isNotBlank()) nowPlaying else "$artist - $title"

                withContext(Dispatchers.Main) {
                    tvTrackTitle.text = displayText

                    if (nextTrack.isNotEmpty()) {
                        tvNextTrackLabel.visibility = View.VISIBLE
                        tvNextTrack.visibility = View.VISIBLE
                        tvNextTrack.text = nextTrack
                    } else {
                        tvNextTrackLabel.visibility = View.GONE
                        tvNextTrack.visibility = View.GONE
                    }

                    val artworkUrl = "https://c7.radioboss.fm/w/artwork/205.jpg?_=" + System.currentTimeMillis()
                    val requestBuilder = Glide.with(this@MainActivity)
                        .load(artworkUrl)
                        .error(R.mipmap.ic_launcher_round)

                    if (ivArtwork.drawable != null) {
                        requestBuilder.placeholder(ivArtwork.drawable)
                    } else {
                        requestBuilder.placeholder(R.mipmap.ic_launcher_round)
                    }

                    requestBuilder.into(ivArtwork)
                }
            }
        }
    }

    private fun isAndroidTV(): Boolean {
        val pm = packageManager
        return pm.hasSystemFeature(PackageManager.FEATURE_LEANBACK) ||
               pm.hasSystemFeature(PackageManager.FEATURE_TELEVISION)
    }

    private fun killAudioAndExit() {
        // Disconnect controller
        mediaController?.let {
            it.stop()
            it.release()
            mediaController = null
        }

        // Stop the background service
        val stopServiceIntent = Intent(this, PlaybackService::class.java)
        stopService(stopServiceIntent)

        // Close app and kill process
        finishAffinity()
        System.exit(0)
    }
}
